model/context.go

```diff
@@ -107,12 +108,42 @@ func truncate(s string) string {
 	return s
 }
 
+type NetworkInfo struct {
+	EffectiveType *string
+	RoundTripTime *int64
+	Downlink      *float64
+	DownlinkMax   *float64
+	SaveData      *bool
+	PhysicalLayer *string
+}
+
+func (ni *NetworkInfo) fields() common.MapStr {
+	if ni == nil {
+		return nil
+	}
+	fields := common.MapStr{}
+	utility.Set(fields, "physical", ni.PhysicalLayer)
+	utility.Set(fields, "downlink", ni.Downlink)
+	utility.Set(fields, "downlink_max", ni.DownlinkMax)
+	utility.Set(fields, "effective_type", ni.EffectiveType)
+	utility.Set(fields, "rtt", ni.RoundTripTime)
+	utility.Set(fields, "saveData", ni.SaveData)
+	return fields
+}
+
 // Page consists of URL and referer
 type Page struct {
 	URL     *URL
 	Referer *string
 }
 
+type Device struct {
+	NetworkInfo            *NetworkInfo
+	Cores                  *int
+	Memory                 *float64
+	ServedViaServiceWorker *string
```

Comment:
Please move `ServedViaServiceWorker` to `Page`. This is not related to the device, and is page-load specific.